tailscale/datasource_devices_test.go

```diff
@@ -0,0 +1,134 @@
+package tailscale_test
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+
+	"github.com/tailscale/terraform-provider-tailscale/tailscale"
+)
+
+func TestAccTailscaleDevices(t *testing.T) {
+	resourceName := "data.tailscale_devices.all_devices"
+
+	// This is a string containing tailscale_device datasource configurations
+	devicesDataSources := &strings.Builder{}
+
+	toResourceComponent := func(str string) string {
+		return strings.ReplaceAll(str, " ", "_")
+	}
+
+	// First test the tailscale_devices datasource, which will give us a list of
+	// all device IDs.
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: testAccProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				Config: `data "tailscale_devices" "all_devices" {}`,
+				Check: func(s *terraform.State) error {
+					client := testAccProvider.Meta().(*tailscale.Clients).V2
+					devices, err := client.Devices().List(context.Background())
+					if err != nil {
+						return fmt.Errorf("unable to list devices: %s", err)
+					}
+
+					devicesByID := make(map[string]map[string]any)
+					for _, device := range devices {
+						m := tailscale.DeviceToMap(&device)
+						m["id"] = device.ID
+						devicesByID[device.ID] = m
+					}
+
+					rs := s.RootModule().Resources[resourceName].Primary
+
+					// first find indexes for devices
+					deviceIndexes := make(map[string]string)
+					for k, v := range rs.Attributes {
+						if strings.HasSuffix(k, ".id") {
+							idx := strings.Split(k, ".")[1]
```

Comment:
nit: For here and below, I always find it very helpful as a reader if detailed string manipulations are accompanied by an example or two in the comments of the expected structure that we're working on.